roachtest: backport improvements to `mixedversion` framework (secure clusters, fixtures opt-out)

renatolabs opened the pull request:
Backports:

  * 3/3 commits from "roachtest: update `mixedversion` to always use secure clusters" (#105454)
  * 1/1 commits from "roachtest: preserve certs in multitenant_tpch" (#105891)
  * 1/1 commits from "roachtest: validate cluster when using fixtures and provide opt-out" (#105455)

Please see individual PRs for details.

/cc @cockroachdb/release

Epic: none

Release note: None

Release justification: test-only changes.

Commit:
roachtest: validate cluster when using fixtures and provide opt-out

This commit updates the `mixedversion` API with two main changes:

* cluster validation: previously, mixed-version tests would always
start from checked-in fixtures. Those fixtures have an implied cluster
topology and require the cluster to have 4 cockroach nodes -- if
that's not the case, the cluster fails to start up with non-obvious
error messages. To make this situation clearer, we now validate that
the test is being run with the correct number of nodes.

* to allow for larger-scale mixed-version tests (more than 4 nodes),
we now expose a `DisableFixtures` option that can be passed to
`NewTest`. With this option, nodes start with an empty store
directory, eliminating any restrictions on cluster size.

Epic: CRDB-19321

Release note: None

Commit:
roachtest: fix bugs related to secure clusters

This commit fixes a number of small bugs related to using secure
clusters in roachtests, which probably stayed dormant up to this point
because we don't use secure clusters enough, especially in tests where
nodes are restarted.

In this commit, we avoid refetching certs when restarting, fix the
checks for certificate existence on nodes, and fix the command to wipe
a cluster when certs are to be preserved.

Epic: none

Release note: None

Commit:
roachtest: introduce `preserveCerts` parameter to `Wipe`

This parameter already existed in the underlying `cluster_synced`
implemenation, but the `cluster` interface exposed to roachtsts forced
that argument to `false`.

This commit changes the interface to expose the parameter, allowing
tests to wipe clusters while keeping the certs intact.

Epic: none

Release note: None

Commit:
roachtest: update `mixedversion` to always use secure clusters

Secure clusters are closer to production deployments and also allow
us to tests features that we couldn't before, like creating new users
with passwords during a test, and then performing SQL operations with
those users.

Epic: CRDB-19321

Release note: None

Commit:
roachtest: preserve certs in multitenant_tpch

The `loadTPCHDataset` function was unconditionally removing certs when
the cluster was wiped. In this commit, we add a parameter to that
function to indicate if the cluster being used is `secure`; if so, we
preserve the certs when wiping.

Fixes: #105880

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106545)
<!-- Reviewable:end -->

renatolabs commented:
Currently running:

`backup-restore/mixed-version` + `multitenant/tpch`: https://teamcity.cockroachdb.com/viewLog.html?buildId=10855325&tab=queuedBuildOverviewTab

`SELECT_PROBABILITY=0.5`: https://teamcity.cockroachdb.com/viewLog.html?buildId=10856169&tab=queuedBuildOverviewTab

Commit:
roachtest: fix `tenant-span-stats/mixed-version`

It needed changes to work with secure clusters, including
authenticating with the server before performing the `curl` call for
the endpoint being tested.

Epic: none

Release note: None

renatolabs commented:
Test passed, merging: https://teamcity.cockroachdb.com/viewLog.html?buildId=10867137&buildTypeId=Cockroach_Nightlies_RoachtestNightlyGceBazel&tab=testsInfo&branch_Cockroach_Nightlies=106545.
